### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by udeaghad**
In this pull request, I have

- Created a class Classroom with the required variables
- Created the has-many/belongs-to relationship between Classroom and Student
- Created a class Book with the required variables
- Created a class Rental with the required variables
- Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental

Kindly review and approve.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rental
```
Comment (BenMukebo) [inline book.rb]: - [ ]  According to the project requirements, you should write two methods to get and set `rentals` and Ruby offers you a nice way to create your getters and setters! you can declare them by using :

```ruby
attr_reader :rentals
attr_writer :rentals
# Or you can make it even shorter by using:
attr_accessor :rentals
```
 Actually, you only have a getter method `attr_reader`. In order to make sure the Getters and Setters for @rentals are present, we need to add a setter method, `attr_writer`.


```ruby
  attr_accessor :title, :author, :rentals
```
Here is a [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor%20#%20*:name%20and%20:age%20instance%20variables%20are%20the%20only%20ones%20that%20should%20be%20able%20to%20be%20modified%20that's%20why%20they%20are%20defined%20as%20attr_accessors%20(here%20both%20Getters%20and%20Setters%20are%20defined)) to the example in a previous lesson: 

Note: You can achieve that by using atrr_reader, attr_writer / attr_accessor, or by writing getters and setters as plain methods ALL OPTIONS are correct as long as getters and setters are created for @rentals

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rental = []
+  end
+
+  def add_rental(rental)
+    @rental.push(rental)
+    rental.book = self
+  end
```
Comment (BenMukebo) [inline book.rb]: - [OPTIONAL] You need to create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`.  Great job creating a method for adding rentals but it should look like this: 

```ruby
  def add_rental(person, date)
    # *the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 
    Rental.new(date, self, person)
  end
```

Here is a [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) to the example in a previous lesson:

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
+
```
Comment (BenMukebo) [inline rental.rb]: - [ ] Same here. Kindly make sure the Getters and Setters for @book and @person instance variables be present 👇🏻 


```ruby
     attr_accessor :date, :book, :person
```

File: person.rb [hunk 4]
```diff
@@ -27,4 +28,9 @@ def can_use_services?
   def correct_name
     @name
   end
+
+  def add_rental(rental)
+    @rental.push(rental)
+    rental.person = self
+  end
```
Comment (BenMukebo) [inline person.rb]: - [OPTIONAL] Great job creating a method for adding rentals  👍🏻 but it should look like this:

```ruby
 def add_rental(book, date)
    # *The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association
    Rental.new(date, book, self)
  end
```

Here is a [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship`) to the example in a previous lesson:

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @udeaghad,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻

- No linter errors. ✔️ 
- Gitflow was correctly used.  ✔️   
- Descriptive README file.  ✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date, book, :person
```
Comment (Bondok6) [inline rental.rb]: - [OPTIONAL] I think you forgot to add simi-colon here, It should be `:book`  instead of `book` 👌

**Review by Bondok6 — APPROVED** [review]
# STATUS: APPROVED :green_circle: :clap: 

## Hi @udeaghad 🙋‍♂️, 
Your project is complete!  There is nothing else to say other than... Great job 👏 
Congratulations! 🎉 

TO HIGHLIGHT:

- Great job on implementing the changes by the previous reviewer ✅
- Your project is professional ✅
- Your repo is professional ✅
- Well structured files ✅
- No Linter errors ✅


Now you can merge this branch and move on :1st_place_medal: 
HAPPY CODING AND GOOD LUCK WITH THE COMMING ONES :crossed_fingers: :four_leaf_clover: 




## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**